ssl/quic/quic_port.c

```diff
@@ -764,39 +936,89 @@ static void port_send_retry(QUIC_PORT *port,
+ * - Token has not expired. Currently set to 10 seconds for tokens from RETRY
+ *   packets and 60 minutes for tokens from NEW_TOKEN packets. This may be
+ *   configurable in the future.
  */
-static int port_validate_token(QUIC_PKT_HDR *hdr, QUIC_CONN_ID *odcid)
+static int port_validate_token(QUIC_PKT_HDR *hdr, QUIC_PORT *port,
+                               BIO_ADDR *peer, QUIC_CONN_ID *odcid,
+                               QUIC_CONN_ID *scid)
 {
-    int valid;
-    struct validation_token *token;
+    int ret = 0;
+    QUIC_VALIDATION_TOKEN token = { 0 };
+    unsigned long long time_diff;
+    size_t remote_addr_len;
+    unsigned char *remote_addr = NULL;
 
-    memset(odcid, 0, sizeof(QUIC_CONN_ID));
+    if (!parse_validation_token(&token, hdr->token, hdr->token_len))
+        goto err;
 
-    token = (struct validation_token *)hdr->token;
-    if (token == NULL || hdr->token_len <= (TOKEN_LEN - QUIC_RETRY_INTEGRITY_TAG_LEN))
-        return 0;
+    /* Validate token timestamp */
+    /* FIXME: Will this handle different timezones, leap seconds, etc? */
+    time_diff = ossl_time2seconds(ossl_time_abs_difference(token.timestamp,
+                                                           ossl_time_now()));
+    if ((token.is_retry && time_diff > 10)
+        || (!token.is_retry && time_diff > 3600))
+        goto err;
 
-    valid = memcmp(token->token_buf, "openssltoken", sizeof("openssltoken") - 1);
-    if (valid != 0)
-        return 0;
+    /* Validate remote address */
+    if (!BIO_ADDR_rawaddress(peer, NULL, &remote_addr_len)
+        || remote_addr_len != token.remote_addr_len
+        || (remote_addr = OPENSSL_malloc(remote_addr_len)) == NULL
+        || !BIO_ADDR_rawaddress(peer, remote_addr, &remote_addr_len)
+        || CRYPTO_memcmp(remote_addr, token.remote_addr, remote_addr_len) != 0)
+        goto err;
 
-    odcid->id_len = token->token_odcid.id_len;
-    memcpy(odcid->id, token->token_odcid.id, token->token_odcid.id_len);
+    /*
+     * Set ODCID and SCID. If the token is from a RETRY packet, retrieve both
+     * from the token. Otherwise, generate a new ODCID and use the header's
+     * source connection ID for SCID.
+     */
+    if (token.is_retry) {
+        /*
+         * We're parsing a packet header before its gone through AEAD validation
+         * here, so there is a chance we are dealing with corrupted data. Make
+         * Sure the dcid encoded in the token matches the headers dcid to
+         * mitigate that.
+         * TODO(QUIC SERVER): Consider handling AEAD validation at the port
+         * level rather than the QRX/channel level to eliminate the need for
+         * this.
+         */
+        if (!memcmp(&token.rscid, &hdr->dst_conn_id, sizeof(QUIC_CONN_ID)))
+            goto err;
+        memcpy(odcid, &token.odcid, sizeof(QUIC_CONN_ID));
+        memcpy(scid, &token.rscid, sizeof(QUIC_CONN_ID));
+    } else {
+        if (!ossl_quic_lcidm_get_unused_cid(port->lcidm, odcid))
```

Comment:
Are you planning on adding NEW_TOKEN frames to this PR, or do we need to open a separate issue for that?

Reply:
There is a separate issue for that: https://github.com/openssl/project/issues/929